packages/nextjs/utils/abi-ninja/proxyContracts.ts

```diff
@@ -0,0 +1,116 @@
+  if (typeof value !== "string" || value === "0x") {
+    throw new Error(`Invalid address value: ${value}`);
+  }
+  const address = value.length === 66 ? "0x" + value.slice(-40) : value;
+  const zeroAddress = "0x" + "0".repeat(40);
+  if (address === zeroAddress) {
+    throw new Error("Empty address");
+  }
+  return address;
+};
+
+const EIP_1167_BYTECODE_PREFIX = "0x363d3d373d3d3d363d";
+const EIP_1167_BYTECODE_SUFFIX = "57fd5bf3";
+
+export const parse1167Bytecode = (bytecode: unknown): string => {
+  if (typeof bytecode !== "string" || !bytecode.startsWith(EIP_1167_BYTECODE_PREFIX)) {
+    throw new Error("Not an EIP-1167 bytecode");
+  }
+
+  // detect length of address (20 bytes non-optimized, 0 < N < 20 bytes for vanity addresses)
+  const pushNHex = bytecode.substring(EIP_1167_BYTECODE_PREFIX.length, EIP_1167_BYTECODE_PREFIX.length + 2);
+  // push1 ... push20 use opcodes 0x60 ... 0x73
+  const addressLength = parseInt(pushNHex, 16) - 0x5f;
+
+  if (addressLength < 1 || addressLength > 20) {
+    throw new Error("Not an EIP-1167 bytecode");
+  }
+
+  const addressFromBytecode = bytecode.substring(
+    EIP_1167_BYTECODE_PREFIX.length + 2,
+    EIP_1167_BYTECODE_PREFIX.length + 2 + addressLength * 2, // address length is in bytes, 2 hex chars make up 1 byte
+  );
+
+  const SUFFIX_OFFSET_FROM_ADDRESS_END = 22;
+  if (
+    !bytecode
+      .substring(EIP_1167_BYTECODE_PREFIX.length + 2 + addressLength * 2 + SUFFIX_OFFSET_FROM_ADDRESS_END)
+      .startsWith(EIP_1167_BYTECODE_SUFFIX)
+  ) {
+    throw new Error("Not an EIP-1167 bytecode");
+  }
+
+  // padStart is needed for vanity addresses
+  return `0x${addressFromBytecode.padStart(40, "0")}`;
+};
+
+export const detectProxyTarget = async (proxyAddress: string, client: any): Promise<string | null> => {
+  try {
+    return await Promise.any([
+      (async () => {
+        const bytecode = await client.getBytecode({ address: proxyAddress });
+        return parse1167Bytecode(bytecode);
+      })(),
+      (async () => {
+        const logicAddress = await client.getStorageAt({ address: proxyAddress, slot: EIP_1967_LOGIC_SLOT });
+        return readAddress(logicAddress);
+      })(),
+      (async () => {
+        const beaconAddress = await client.getStorageAt({ address: proxyAddress, slot: EIP_1967_BEACON_SLOT });
+        const resolvedBeaconAddress = readAddress(beaconAddress);
+        for (const method of EIP_1167_BEACON_METHODS) {
+          try {
+            const data = await client.call({ data: method as `0x${string}`, to: resolvedBeaconAddress });
+            return readAddress(data.data);
+          } catch {
+            // Ignore
+          }
+        }
+        throw new Error("Beacon method calls failed");
+      })(),
+    ]);
+  } catch (error) {
+    try {
+      return await Promise.any([
+        (async () => {
+          const logicAddress = await client.getStorageAt({ address: proxyAddress, slot: EIP_1822_LOGIC_SLOT });
```

Comment:
Any particular reason to break from the first Promise.any and then again continue in catch for other methods 

Wouldn't it have been possible to batch this methods in `catch` of Promise.any inside Promise.any of `try ` block keeping only one Promise.any ? 

> Adds the ability to detect Zora contracts by first checking for EIP-1967

Or this is the reason we separated them in two `Promise.any` so that we always make sure `EIP_1967` logic is executed first (keeping it in first Promise.any ) 

Because I think when we use Promise.any() request are fired parellely and  we never know which is gonna execute first. 

Really sorry for dumb question and lol not sure if querstion makes sense

Reply:
Oh don't be sorry at all, it is my bad not explaining it in the first post! 
Yes, the reason is the Zora contract and other contracts that are similar to it. When we try to resolve all requests at once, there is like 1/2 chance of the contract getting detected in the wrong way. 
So what we are doing is that we first check for 1967 and only if it fails we check for 897.

> Or this is the reason we separated them in two Promise.any so that we always make sure EIP_1967 logic is executed first (keeping it in first Promise.any )

Yes. This is the exact reason.

(somehow i drafted this but never clicked comment... Sorry about this. but yes)